### ISSUE TITLE
Implement expire time for DKIM signatures

### ISSUE THREAD

**Issue by ghost** [rspamd/rspamd#2127]
### Classification (Please choose *one* option):

* [ ] Crash/Hang/Data loss
* [ ] WebUI/Usability
* [ ] Serious bug
* [ ] Other bug
* [X] Feature
* [ ] Enhancement

### Reproducibility (Please choose *one* option):

* [ ] Always
* [ ] Sometimes
* [ ] Rarely
* [ ] Unable
* [ ] I didn’t try
* [X] Not applicable

### Rspamd version:
Tested with 1.6.6 (this is what HardenedBSD currently ships), but this is also applicable to 1.7.0+

### Operation system, CPU, memory and environment:
All

### Description (Please provide a descriptive summary of the issue):
Implement an option to set an expire timestamp for DKIM signatures. Some people rotate their DKIM keys reguarly, and it does not make sense to create signatures that can be still validated in a long time.

This feature is implemented in OpenDKIM and documented here: http://www.opendkim.org/opendkim.conf.5.html (section: SignatureTTL).

### Additional information:
Thank you very much.

**Comment by heraklit256** [comment]
Update: Also applicable for 1.7.2. :-)

**Comment by vstakhov** [comment]
After some more thinking I've concluded that this feature is absolutely useless. If you want to expire old signature, you can just change a selector and remove an old public key eventually. The meaning of signature expiration seems to be bogus for me.

**Comment by heraklit256** [comment]
Sorry to diagree here.

First: In case a DKIM key has been rotated and an existing signature cannot be validated anymore, validation simply failes. If the signature has already been expired, you get a completely different result (assuming validaton failure != expired signature). This enables you to rotate DKIM keys, guaranteed that no one will ever run into validation failures (publish a new key, use it for signing, and wait until expire time of last message signed with old key is reached, then old DKIM key can be safely removed).

Second: A DKIM key proves a message has been relayed through some of your systems. Maybe someone does not want this forever, but just for the time a mail is typically delivered (5d, usually 14 or 30 days to be sure).

To sum it up: In case one would (ab)use this as a mechanism for key rotating, you are right. Otherwise, I would prefer to have limited lifetime on DKIM signatures.

Maybe @andreasschulze can be more precise here than I do. :-)

**Comment by andreasschulze** [comment]
An expired signature has an other semantic than a messages that cannot be validated because of the missing public key. Yes, there is no huge difference. I see it more as an explicit statement about validity times.

**Comment by vstakhov** [comment]
Expired signatures are completely ignored, missing of DNS key causes `DKIM_TEMPFAIL`, empty key (as suggested by DKIM guidelines) will produce `DKIM_PERMFAIL`.

**Comment by patrickbenkoetter** [comment]
> After some more thinking I've concluded that this feature is absolutely useless. If you want to expire old signature, you can just change a selector and remove an old public key eventually. The meaning of signature expiration seems to be bogus for me.

Please have a look at DKIM Replay Attacks and maybe reconsider your verdict. Setting an expiration time makes platforms that are being abused less attractive for miscreants.

**Comment by vstakhov** [comment]
The threat model is still not clear to me. What is the purpose of DKIM replay attack counting that you cannot change the message without invalidating the signature? The only thing that comes to my mind is double-spending alike attacks, but are they relevant to email?

**Comment by patrickbenkoetter** [comment]
> The threat model is still not clear to me. What is the purpose of DKIM replay attack counting that you cannot change the message without invalidating the signature? The only thing that comes to my mind is double-spending alike attacks, but are they relevant to email?

This is how a DKIM replay attack works:

1. An attacker uses a platforms good reputation to send a test message, which contains the spam message. The message passes filters and gets delivered to the attackers mailbox. The attacker now has a spam message with a valid DKIM signature
2. The attacker uses that message as blueprint and sends it out to all victims leaving the message intact alternating only the envelope recipient per victim

Setting an expiration time has proven to be an effective measure to limit this type of abuse. Of course there are other techniques like requiring strict alignment that also work. It's just not feasible in cases that are still being used.

**Comment by vstakhov** [comment]
I understand that, however, the proposed sollution to implement `t=` support seems to be pointless to fix this issue. One can set frequent keys/selectors rotation and stop publishing old keys with clearly the same effect as `t=` property. The only difference is that the signatures will emit `DKIM_PERMFAIL` instead of `DKIM_REJECT` but semantically they are the same - a verifier must treat both rejected and permfailed (e.g. due to DNS failure) records equally as invalid/absent signatures.

Apparently, both setting expire and keys rotation does not fix replay issue as there is still a window for replay attack , and the size of this window must be large enough to avoid deliverability issues.

**Comment by vstakhov** [comment]
BTW, from RFC 6376: `INFORMATIVE NOTE: The "x=" tag is not intended as an antireplay defense.`

**Comment by vstakhov** [comment]
And the key revocation is advised as a workaround for that, specifically in [8.7](https://datatracker.ietf.org/doc/html/rfc6376#section-8.7).

**Comment by patrickbenkoetter** [comment]
> BTW, from RFC 6376: `INFORMATIVE NOTE: The "x=" tag is not intended as an antireplay defense.`

Shoot! In this case forget my request and sorry for the noise. I had catched this up from a panel member as something "that worked" and didn't do my homework first.
